### PR TITLE
Issue #11 Remove items from shopping list

### PR DESCRIPTION
**Pull request by katia-utochkina**
## Description

This PR allows a user to delete items from their shopping list so that their list is not cluttered with items they don't want to buy in the future. To prevent accidental deletions, it asks users to confirm the deletion of the items.The PR ensures that removing items from the app UI removes them from the Firestore db, too.

## Acceptance Criteria

- [x] User is able to delete an item from the shopping list
- [x] Before deleting, prompt the user to confirm that they really want to delete the item to prevent accidental deletions
- [x] Deletion should cause the associated record(s) in the database to be deleted

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Testing Steps / QA Criteria

* From your terminal, pull down this branch with `git pull origin ku-ci-remove-items-from-shopping-list` and check that branch out with `git checkout ku-ci-remove-items-from-shopping-list` 
* Then `npm start` to launch the app
* Check that ``Remove`` button is there
* When clicking on the ``Remove`` button, check that you are asked to either confirm or cancel the deletion
* Make sure the item is deleted both off the list and the Firestore db
* Make sure that the item remains in the list if you cancel the action

### CODE REVIEW

**Review by lamilekan263 — APPROVED** [review]

**Review by C-Zekeri — COMMENTED** [review]
Well done Katia and Cynthia, you've done a great job! Picked up a few things from reading this, thank you!

Noticed something while testing though:
With the new change you made to the regex, if I add "Toilet Paper" to the list, and then try adding "toilet paper", toilet paper gets converted to Toilet Paper, and now we have two instances of Toilet Paper on the list.

Everything else works as expected.

**Comment by katia-utochkina** [comment]
Hi @C-Zekeri , the issue is fixed

**Review by C-Zekeri — APPROVED** [review]

File: src/App.css [hunk 1]
```diff
@@ -52,22 +52,39 @@ h1 {
...
+  animation: bouncing-loader 0.6s infinite alternate;
+}
+
+.bouncing-loader > div:nth-child(2) {
+  animation-delay: 0.2s;
+}
+
+.bouncing-loader > div:nth-child(3) {
+  animation-delay: 0.4s;
 }
```
Comment (faysvas) [inline src/App.css:82]: Adding a loader is a nice touch and gives feedback to the user :+1: !

File: src/lib/utils.js [hunk 2]
```diff
@@ -1,4 +1,7 @@
 const transformUserInput = (item) =>
-  item.replace(/[\W_]/g, '').toLowerCase().trim();
+  item
+    .replace(/[^a-zA-Z ]/g, '')
+    .replace(/\s\s+/g, ' ')
+    .toLowerCase();
```
Comment (faysvas) [inline src/lib/utils.js:6]: What did you change here?

File: src/components/BinIcon.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    <path
+      style={{ fill: '000' }}
+      d="m20.651 2.3339c-.73869 0-1.3312.59326-1.3312 1.3296v2.5177h-6.3634c-.73887 0-1.3314.59331-1.3314 1.3295v1.1888c0 .73639.59249 1.3289 1.3312 1.3289h7.6948 8.8798 7.6948c.73869 0 1.3312-.59249 1.3312-1.3289v-1.1888c0-.73639-.59249-1.3296-1.3312-1.3296h-6.3634v-2.5177c0-.73639-.59249-1.3296-1.3312-1.3296h-8.8798zm-5.6786 11.897c-1.7775 0-3.2704 1.4889-3.2704 3.274v27.647c0 1.7775 1.4928 3.2704 3.2704 3.2704h20.783c1.7775 0 3.2704-1.4928 3.2704-3.2704v-27.647c0-1.7852-1.4928-3.274-3.2704-3.274h-20.783zm1.839 3.4895h1.1696c.73869 0 1.3389.60018 1.3389 1.3466v24.247c0 .74638-.60018 1.3389-1.3389 1.3389h-1.1696c-.73869 0-1.3389-.59249-1.3389-1.3389v-24.247c0-.74638.60018-1.3466 1.3389-1.3466zm7.6948 0h1.1696c.73869 0 1.3389.60018 1.3389 1.3466v24.247c0 .74638-.60018 1.3389-1.3389 1.3389h-1.1696c-.73869 0-1.3389-.59249-1.3389-1.3389v-24.247c0-.74638.60018-1.3466 1.3389-1.3466zm7.6948 0h1.1696c.73869 0 1.3389.60018 1.3389 1.3466v24.247c0 .74638-.60018 1.3389-1.3389 1.3389h-1.1696c-.73869 0-1.3389-.59249-1.3389-1.3389v-24.247c0-.74638.60018-1.3466 1.3389-1.3466z"
+    ></path>
+  </svg>
+);
+
+export default BinIcon;
```
Comment (faysvas) [inline src/components/BinIcon.js:12]: I like that you made a separate component for the bin icon. This way, you avoid cluttering the ListItem component code with svg details :100:

File: src/App.css [hunk 4]
```diff
@@ -250,16 +287,76 @@ nav {
...
+  display: inline-flex;
+  align-items: center;
   justify-content: center;
+  flex-shrink: 0;
+  width: 4.5ch;
+  height: 4.5ch;
+  border-radius: none;
+  background: none;
+  border: none;
+  fill: var(--theme-color);
```
Comment (faysvas) [inline src/App.css]: I would suggest you add a `cursor: pointer;` here to make it more obvious that the trash can is a button :slightly_smiling_face:

**Review by faysvas — APPROVED** [review]
Nice job! I like how you structured the code :1st_place_medal:. I asked a question and made a suggestion for some CSS but you can consider this approved.

PS
I didn't review the CSS because I assumed it's work in progress. I'll review the entire CSS once the functionality of the app is done :slightly_smiling_face:

File: src/lib/utils.js (hunk 2 above)
Comment (katia-utochkina) [reply src/lib/utils.js:6]: Hi @faysvas . because you said there is a "Bug: Spaces between multi-word items disappear"?

File: src/App.css (hunk 4 above)
Comment (katia-utochkina) [reply src/App.css]: Oh yes, you're right. Just added this

File: src/lib/utils.js (hunk 2 above)
Comment (ciradu2204) [reply src/lib/utils.js]: Hi @faysvas , when I try to use the regex we had before, it works perfectly on my side. Not sure what was the problem. 

image

File: src/App.css (hunk 1 above)
Comment (YuriDevAT) [reply src/App.css:82]: Totally, I was very happy to see this feature :)

File: src/App.css [hunk 5]
```diff
@@ -151,6 +163,35 @@ input:-webkit-autofill:active {
   cursor: pointer;
 }
 
+/*--------------------------------------------------------------------------*\
+     LIST VIEW
+\*--------------------------------------------------------------------------*/
```
Comment (YuriDevAT) [inline src/App.css:168]: Separating the CSS of the different views makes this code much more readable, good choice to go with comments 👍🏽

File: src/components/BinIcon.js (hunk 3 above)
Comment (YuriDevAT) [reply src/components/BinIcon.js:12]: 💯  agree

File: src/lib/utils.js (hunk 2 above)
Comment (YuriDevAT) [reply src/lib/utils.js:6]: So, this PR also closes the new issue made opened by Golfo a few days ago, great 👍🏽

File: src/App.css [hunk 6]
```diff
@@ -250,16 +287,75 @@ nav {
...
+  border-radius: none;
+  background: none;
+  border: none;
+  cursor: pointer;
+}
+
+.list__item__delete-btn svg {
+  margin-bottom: 10px;
+}
+
```
Comment (YuriDevAT) [inline src/App.css:311]: Something here seems not right when I am in my Desktop view using Firefox. The bins are not in one line (and I guess this is wanted). Also, when adding a very long item, the view seems a bit off.
Screenshot 2021-05-13 at 20 39 13

**Review by YuriDevAT — APPROVED** [review]
The way you have written your code seems sooo professional, very well done team. I am already very excited about your demos on Sunday 😁  A left a few comments/suggestions! I already approved your code since the functionality works as it should and you fulfill the AC.

**Pull request merged**